### ISSUE TITLE
Study the threading efficiency of Fall 2016 re-RECO jobs

### ISSUE THREAD

**Issue by Dr15Jones** [cms-sw/cmssw#16104]
The Fall 2016 re-RECO jobs are showing signs of in-efficiencies:  https://hypernews.cern.ch/HyperNews/CMS/get/edmFramework/3683.html

We should look at a representative job and study its ratio of serial to parallel processing.

**Comment by Dr15Jones** [comment]
@bbockelm

**Comment by Dr15Jones** [comment]
assign core

**Comment by slava77** [comment]
was the ROOT patch for LZMA from Dan integrated in the 80X rereco release?
This should be 5-10% of inefficiency

**Comment by bbockelm** [comment]
Do not know.  The release is 8_0_20; sample pset is at `~bbockelm/public/logs/sample_rereco_job.tar.gz` (CERN AFS) if you'd like to try running the job for yourself.

**Comment by Dr15Jones** [comment]
From the tracer log we can also determine where non-event processing time is spent not processing events. Below are tracer output for major framework transitions
```
05-Oct-2016 17:38:20.16 CEST  ++ starting: constructing source: PoolSource
05-Oct-2016 17:38:32.71 CEST  ++ finished: constructing source: PoolSource
05-Oct-2016 17:38:32.71 CEST  ++++ starting: constructing module with label 'TriggerResults' id = 1
...
05-Oct-2016 17:39:49.50 CEST  ++++ finished: constructing module with label 'zdcreco' id = 2076
05-Oct-2016 17:39:51.95 CEST  ++ preallocate: 1 concurrent runs, 1 concurrent luminosity sections, 4 streams
05-Oct-2016 17:39:51.97 CEST  ++ starting: begin job
...
05-Oct-2016 17:39:52.28 CEST  ++ finished: begin job
05-Oct-2016 17:39:52.28 CEST  ++++ starting: begin stream for module: stream = 0 label = 'jetsForCoreTrackingPreSplitting' id = 2
...
05-Oct-2016 17:39:53.26 CEST  ++++ finished: begin stream for module: stream = 3 label = 'zdcreco' id = 2076
05-Oct-2016 17:39:54.07 CEST  ++++ starting: source run
05-Oct-2016 17:39:54.07 CEST  ++++ finished: source run
05-Oct-2016 17:39:54.94 CEST  ++++ starting: global begin run 276242 : time = 6302860503971310080
05-Oct-2016 17:39:58.23 CEST  ++++ finished: global begin run 276242 : time = 6302860503971310080
05-Oct-2016 17:39:58.23 CEST  ++++ starting: begin run: stream = 0 run = 276242 time = 6302860503971310080
05-Oct-2016 17:40:35.91 CEST  ++++ finished: begin run: stream = 0 run = 276242 time = 6302860503971310080
05-Oct-2016 17:40:35.91 CEST  ++++ starting: begin run: stream = 1 run = 276242 time = 6302860503971310080
05-Oct-2016 17:40:57.32 CEST  ++++ finished: begin run: stream = 1 run = 276242 time = 6302860503971310080
05-Oct-2016 17:40:57.32 CEST  ++++ starting: begin run: stream = 2 run = 276242 time = 6302860503971310080
05-Oct-2016 17:41:18.41 CEST  ++++ finished: begin run: stream = 2 run = 276242 time = 6302860503971310080
05-Oct-2016 17:41:18.41 CEST  ++++ starting: begin run: stream = 3 run = 276242 time = 6302860503971310080
05-Oct-2016 17:41:39.14 CEST  ++++ finished: begin run: stream = 3 run = 276242 time = 6302860503971310080
05-Oct-2016 17:41:39.29 CEST  ++++ starting: global begin lumi: run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:39.53 CEST  ++++ finished: global begin lumi: run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:39.53 CEST  ++++ starting: begin lumi: stream = 0 run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:40.06 CEST  ++++ finished: begin lumi: stream = 0 run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:40.06 CEST  ++++ starting: begin lumi: stream = 1 run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:40.34 CEST  ++++ finished: begin lumi: stream = 1 run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:40.34 CEST  ++++ starting: begin lumi: stream = 2 run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:40.61 CEST  ++++ finished: begin lumi: stream = 2 run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:40.61 CEST  ++++ starting: begin lumi: stream = 3 run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:40.87 CEST  ++++ finished: begin lumi: stream = 3 run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 17:41:44.68 CEST  ++++ starting: processing event : stream = 0 run = 276242 lumi = 1319 event = 2474550263 time = 6302860503971310080
...
05-Oct-2016 18:29:11.42 CEST  ++++ finished: processing event : stream = 1 run = 276242 lumi = 1319 event = 2473915193 time = 6302860469610914048
05-Oct-2016 18:29:11.51 CEST  ++++ starting: end lumi: stream = 0 run = 276242 lumi = 1319 time = 6302860503971201024
05-Oct-2016 18:29:11.83 CEST  ++++ finished: end lumi: stream = 0 run = 276242 lumi = 1319 time = 6302860503971201024
05-Oct-2016 18:29:11.83 CEST  ++++ starting: end lumi: stream = 1 run = 276242 lumi = 1319 time = 6302860503971201024
05-Oct-2016 18:29:12.11 CEST  ++++ finished: end lumi: stream = 1 run = 276242 lumi = 1319 time = 6302860503971201024
05-Oct-2016 18:29:12.11 CEST  ++++ starting: end lumi: stream = 2 run = 276242 lumi = 1319 time = 6302860503971201024
05-Oct-2016 18:29:12.39 CEST  ++++ finished: end lumi: stream = 2 run = 276242 lumi = 1319 time = 6302860503971201024
05-Oct-2016 18:29:12.39 CEST  ++++ starting: end lumi: stream = 3 run = 276242 lumi = 1319 time = 6302860503971201024
05-Oct-2016 18:29:12.68 CEST  ++++ finished: end lumi: stream = 3 run = 276242 lumi = 1319 time = 6302860503971201024
05-Oct-2016 18:29:12.68 CEST  ++++ starting: global end lumi: run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 18:29:15.63 CEST  ++++ finished: global end lumi: run = 276242 lumi = 1319 time = 6302860503971310080
05-Oct-2016 18:29:15.76 CEST  ++++ starting: end run: stream = 0 run = 276242 time = 6302860503971201024
05-Oct-2016 18:29:17.04 CEST  ++++ finished: end run: stream = 0 run = 276242 time = 6302860503971201024
05-Oct-2016 18:29:17.04 CEST  ++++ starting: end run: stream = 1 run = 276242 time = 6302860503971201024
05-Oct-2016 18:29:17.62 CEST  ++++ finished: end run: stream = 1 run = 276242 time = 6302860503971201024
05-Oct-2016 18:29:17.62 CEST  ++++ starting: end run: stream = 2 run = 276242 time = 6302860503971201024
05-Oct-2016 18:29:18.19 CEST  ++++ finished: end run: stream = 2 run = 276242 time = 6302860503971201024
05-Oct-2016 18:29:18.19 CEST  ++++ starting: end run: stream = 3 run = 276242 time = 6302860503971201024
05-Oct-2016 18:29:18.76 CEST  ++++ finished: end run: stream = 3 run = 276242 time = 6302860503971201024
05-Oct-2016 18:29:18.76 CEST  ++++ starting: global end run 276242 : time = 6302860503971201024
05-Oct-2016 18:29:18.99 CEST  ++++ finished: global end run 276242 : time = 6302860503971201024
05-Oct-2016 18:29:21.81 CEST  ++++ starting: close input file: lfn = /store/data/Run2016C/SingleElectron/RAW/v2/000/276/242/00001/DCE914D3-A640-E611-B9F6-02163E014429.root
05-Oct-2016 18:29:21 CEST  Closed file root://xrootd-cms.infn.it//store/data/Run2016C/SingleElectron/RAW/v2/000/276/242/00001/DCE914D3-A640-E611-B9F6-02163E014429.root
05-Oct-2016 18:29:21.82 CEST  ++++ finished: close input file: lfn = /store/data/Run2016C/SingleElectron/RAW/v2/000/276/242/00001/DCE914D3-A640-E611-B9F6-02163E014429.root
05-Oct-2016 18:32:06.45 CEST  ++++ starting: end stream for module: stream = 0 label = 'jetsForCoreTrackingPreSplitting' id = 2
...
05-Oct-2016 18:32:07.41 CEST  ++++ finished: end stream for module: stream = 3 label = 'zdcreco' id = 2076
05-Oct-2016 18:32:07.41 CEST  ++++ starting: end job for module with label 'APVPhases' id = 808
...
05-Oct-2016 18:32:07.63 CEST  ++ finished: end job
```

**Comment by Dr15Jones** [comment]
The plot of what each stream is doing for this job is quite large and github won't accept it. If you wish to view it you can get it from https://dl.dropboxusercontent.com/u/11356841/rereco_1000_stall.pdf

**Comment by bbockelm** [comment]
Chris - I'm having a hard time digesting the raw data above about framework transitions.  Can you summarize it for me?

**Comment by Dr15Jones** [comment]
From the raw tracer output, the time in transitions are

1. construction: 89.34s
1. begin job: 0.31s
1. begin stream: 0.98s
1. begin run: 105.07s
1. begin lumi: 1.58s
1. event processing: 2846.74s
1. end lumi: 4.12s
1. end run: 3.23s
1. between source closing and begin of end stream (output module closing?): 164.63s
1. end stream: 0.96s
1. end job: 0.22

**Comment by Dr15Jones** [comment]
The timing report for the job has
```
TimeReport> Time report complete in 3135.35 seconds
 Time Summary: 
 - Min event:   2.92212
 - Max event:   107.752
 - Avg event:   11.0727
 - Total loop:  2966.71
 - Total job:   3135.35
 Event Throughput: 0.337074 ev/s
 CPU Summary: 
 - Total loop:  9739.11
 - Total job:   9906.68
```

**Comment by Dr15Jones** [comment]
From the timing report, we can estimate our speedup as
`9739.11/2966.71 = 3.3`
or efficiency of `3.3/4 = 0.825`

**Comment by bbockelm** [comment]
That approximately matches what we see in production.

**Comment by Dr15Jones** [comment]
From the transition timing, if the event processing was 100% thread efficient,  then the contribution to the CPU inefficiency by the non-event parts would be
```
serial time = 89.34+0.31+0.98+105.07+1.58+4.12+3.23+164.63+0.96+0.22 = 370.44
efficiency = (370.44 + 4*2846.74)/(370.44+2846.74)/4 = 0.913641760797
```

but instead of assuming perfect event processing, we can use the values from the time report
```
efficiency = (370.44 + 9739.11)/(370.44+2846.74)/4 = 0.785590952325
```
which compares to the just event loop efficiency of `0.825`

**Comment by Dr15Jones** [comment]
I ran the exact same job, but this time I first copied the input file to the local disk. The stall information for this job is
```
Stalled Module                       Tot Stall Time  Stall Times
source                              : 1478.98        92.44, 92.00, 91.67, 90.38, 87.89, 87.02, 84.03, 83.58, 83.34, 83.27, 81.97, 75.69, 27.75, 25.47, 24.89, 23.39, 22.53, 17.89, 16.67, 16.56, 16.11, 15.85, 15.48, 13.26, 13.07, 12.82, 11.89, 9.07, 8.59, 8.33, 7.38, 6.82, 3.61, 3.58, 2.30, 2.29, 2.13, 2.05, 1.92, 1.74, 1.67, 1.52, 1.48, 1.47, 1.40, 1.16, 0.86, 0.79, 0.64, 0.63, 0.62, 0.62, 0.61, 0.54, 0.51, 0.51, 0.50, 0.46, 0.44, 0.41, 0.40, 0.37, 0.33, 0.31, 0.29, 0.25, 0.25, 0.25, 0.25, 0.25, 0.25, 0.24, 0.22, 0.21, 0.20, 0.20, 0.20, 0.20, 0.20, 0.20, 0.20, 0.20, 0.20, 0.20, 0.20, 0.20, 0.20, 0.19, 0.18, 0.18, 0.18, 0.18, 0.18, 0.18, 0.18, 0.17, 0.17, 0.17, 0.17, 0.17, 0.17, 0.17, 0.17, 0.17, 0.17, 0.17, 0.17, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.15, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.14, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10
AODoutput                           : 1386.80        92.43, 92.00, 91.66, 90.37, 87.88, 87.02, 84.03, 83.58, 83.33, 83.26, 81.96, 75.68, 27.75, 25.47, 24.88, 23.38, 22.52, 17.88, 16.67, 16.55, 16.10, 15.85, 15.47, 13.26, 13.06, 12.81, 11.89, 9.07, 8.59, 8.32, 7.37, 6.81, 3.60, 3.58, 2.29, 2.28, 2.13, 2.04, 1.91, 1.73, 1.67, 1.52, 1.48, 1.47, 1.39, 1.15, 0.86, 0.78, 0.63, 0.63, 0.61, 0.61, 0.61, 0.53, 0.50, 0.50, 0.50, 0.46, 0.44, 0.41, 0.39, 0.37, 0.30, 0.28, 0.23, 0.21, 0.21, 0.20, 0.19, 0.18, 0.16, 0.15, 0.14, 0.14, 0.13, 0.11, 0.10, 0.10
MINIAODoutput                       : 97.77          37.43, 36.66, 19.34, 0.46, 0.42, 0.38, 0.38, 0.33, 0.27, 0.26, 0.21, 0.20, 0.20, 0.18, 0.17, 0.17, 0.16, 0.16, 0.16, 0.12, 0.11
valHcalTriggerPrimitiveDigis        : 5.30           1.79, 1.76, 1.75
l1tCsctf                            : 4.62           1.57, 1.57, 1.48
ALCARECOStreamHcalCalIterativePhiSym: 1.53           0.85, 0.68
ALCARECOStreamEcalESAlign           : 1.22           0.49, 0.49, 0.24
ckfOutInTracksFromConversions       : 0.51           0.51
pfJetProbabilityBJetTagsPuppi       : 0.33           0.33
tightinclusiveCandidateVertexFinder : 0.32           0.32
valRctDigis                         : 0.29           0.15, 0.14
hcalLaserEventFilter                : 0.19           0.19
patPFMetT1UnclusteredEnDown         : 0.19           0.19
valGctDigis                         : 0.14           0.14
fixedGridRhoFastjetAllTmp           : 0.14           0.14
firstStepPrimaryVertices            : 0.14           0.14
offlinePrimaryVertices              : 0.12           0.12
ALCARECOStreamEcalUncalWElectron    : 0.11           0.11
siPixelDigis                        : 0.10           0.10
```

and the timing summary is
```
TimeReport> Time report complete in 3102.44 seconds
 Time Summary: 
 - Min event:   3.13513
 - Max event:   115.668
 - Avg event:   11.2562
 - Total loop:  2957.21
 - Total job:   3102.44
 Event Throughput: 0.338157 ev/s
 CPU Summary: 
 - Total loop:  9871.98
 - Total job:   10016.1
```
Which gives an efficiency of
```9871.98/2957.21/4 = 0.834568732014```

**Comment by Dr15Jones** [comment]
Looking at the log files, the edmStreamStallGrapher appears to be incorrectly assigning time to the Source. The actual stalls appear to be coming from the AODoutput PoolOutputModule.

**Comment by Dr15Jones** [comment]
@dan131riley it would help having another set of eyes on this problem.

**Comment by Dr15Jones** [comment]
From the local file read, I find the following tracer information
```
05-Oct-2016 21:25:38.53 CEST  ++++++++ starting: processing event for module: stream = 3 label = 'AODoutput' id = 800
05-Oct-2016 21:25:38.53 CEST  ++++++++ starting: delayed processing event for module: stream = 3 label = 'AODoutput' id = 800
05-Oct-2016 21:25:38.53 CEST  ++++++++ finished: delayed processing event for module: stream = 3 label = 'AODoutput' id = 800
...
05-Oct-2016 21:25:46.39 CEST  ++++++++ finished: delayed processing event for module: stream = 2 label = 'AODoutput' id = 800
05-Oct-2016 21:25:46.85 CEST  ++++++++ finished: delayed processing event for module: stream = 0 label = 'AODoutput' id = 800
05-Oct-2016 21:25:47.21 CEST  ++++++++ finished: delayed processing event for module: stream = 1 label = 'AODoutput' id = 800
[big time gap in log file]
05-Oct-2016 21:27:18.82 CEST  ++++++++ finished: processing event for module: stream = 3 label = 'AODoutput' id = 800
...
05-Oct-2016 21:27:18.82 CEST  ++++++++ starting: processing event for module: stream = 2 label = 'AODoutput' id = 800
05-Oct-2016 21:27:18.85 CEST  ++++++++ starting: processing event for module: stream = 0 label = 'AODoutput' id = 800
05-Oct-2016 21:27:18.87 CEST  ++++++++ finished: processing event for module: stream = 0 label = 'AODoutput' id = 800
05-Oct-2016 21:27:18.87 CEST  ++++++++ starting: processing event for module: stream = 1 label = 'AODoutput' id = 800
```
So what is happening is stream 3 starts running the AODoutput module and takes more than 90 seconds. The other streams eventually get to the point that they need to run the AODoutput module and they all are stuck waiting. The stall graph (which I will post) shows similarly long waits occur 4 times in this job.

**Comment by Dr15Jones** [comment]
https://dl.dropboxusercontent.com/u/11356841/rereco_stall_local_1000.pdf

**Comment by Dr15Jones** [comment]
The part of the job which has the initial long time in the output module has the following events to stream processing
```
05-Oct-2016 21:25:24.39 CEST  ++++ starting: processing event : stream = 3 run = 276242 lumi = 1319 event = 2474402544 time = 6302860495381370880
05-Oct-2016 21:25:26.14 CEST  ++++ finished: processing event : stream = 1 run = 276242 lumi = 1319 event = 2474129234 time = 6302860482495733504
05-Oct-2016 21:25:26.21 CEST  ++++ starting: processing event : stream = 1 run = 276242 lumi = 1319 event = 2475241190 time = 6302860546920773120
05-Oct-2016 21:25:30.23 CEST  ++++ finished: processing event : stream = 0 run = 276242 lumi = 1319 event = 2474561404 time = 6302860503971463424
05-Oct-2016 21:25:30.30 CEST  ++++ starting: processing event : stream = 0 run = 276242 lumi = 1319 event = 2474605778 time = 6302860508266034045
05-Oct-2016 21:25:32.05 CEST  ++++ finished: processing event : stream = 1 run = 276242 lumi = 1319 event = 2475241190 time = 6302860546920773120
05-Oct-2016 21:25:32.11 CEST  ++++ starting: processing event : stream = 1 run = 276242 lumi = 1319 event = 2474605794 time = 6302860508266034176
05-Oct-2016 21:25:34.48 CEST  ++++ finished: processing event : stream = 2 run = 276242 lumi = 1319 event = 2473858282 time = 6302860465316166912
05-Oct-2016 21:25:34.57 CEST  ++++ starting: processing event : stream = 2 run = 276242 lumi = 1319 event = 2474477312 time = 6302860499676353024
05-Oct-2016 21:25:37.66 CEST  ++++ finished: processing event : stream = 0 run = 276242 lumi = 1319 event = 2474605778 time = 6302860508266034045
05-Oct-2016 21:25:37.72 CEST  ++++ starting: processing event : stream = 0 run = 276242 lumi = 1319 event = 2473989083 time = 6302860473905883136
05-Oct-2016 21:25:39.09 CEST  ++++ finished: processing event : stream = 1 run = 276242 lumi = 1319 event = 2474605794 time = 6302860508266034176
05-Oct-2016 21:25:39.14 CEST  ++++ starting: processing event : stream = 1 run = 276242 lumi = 1319 event = 2475325291 time = 6302860551215983360
05-Oct-2016 21:27:20.05 CEST  ++++ finished: processing event : stream = 2 run = 276242 lumi = 1319 event = 2474477312 time = 6302860499676353024
05-Oct-2016 21:27:20.06 CEST  ++++ finished: processing event : stream = 3 run = 276242 lumi = 1319 event = 2474402544 time = 6302860495381370880
05-Oct-2016 21:27:20.06 CEST  ++++ finished: processing event : stream = 0 run = 276242 lumi = 1319 event = 2473989083 time = 6302860473905883136
05-Oct-2016 21:27:20.08 CEST  ++++ finished: processing event : stream = 1 run = 276242 lumi = 1319 event = 2475325291 time = 6302860551215983360
```

**Comment by dan131riley** [comment]
8_0_20 does not have the LZMA patch:

```
dr13_lnx231% echo $CMSSW_RELEASE_BASE 
/cvmfs/cms.cern.ch/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_0_20
dr13_lnx231% nm $ROOTSYS/lib/libCore.so | fgrep lzma
                 U lzma_code@@XZ_5.0
                 U lzma_easy_encoder@@XZ_5.0
                 U lzma_end@@XZ_5.0
                 U lzma_stream_decoder@@XZ_5.0
```

**Comment by Dr15Jones** [comment]
Do we know what pre-release the LZMA patch was first introduced? [Assuming it is in 8_1_X.]

**Comment by bbockelm** [comment]
Some output notes:

- All the objects are already serialized.  When we hit event "cluster" (ROOT terminology) boundaries, every active basket (so, 5,000 baskets) is compressed and written to disk.
- Here's the for-loop where we iterate over all top-level branches and flush each: https://github.com/root-mirror/root/blob/master/tree/tree/src/TBranch.cxx#L2602
- Digging through the call graph, it seems that `TBranch::FlushBaskets` is mostly thread-safe: we could do multiple branches at once as part of a TBB parallel-for loop.
- The only non-thread-safe construct I found in the Tree/Branch layer is updating some statistics: https://github.com/root-mirror/root/blob/master/tree/tree/src/TBranch.cxx#L2602
- In the TBasket code itself, there are plenty of non-thread-safe constructs when interacting with the file (https://github.com/root-mirror/root/blob/master/tree/tree/src/TBasket.cxx#L1001).  Should be possible to have a mutex managing the interaction with TFile - but allowing the compression to proceed in parallel.

Thoughts?

**Comment by Dr15Jones** [comment]
@davidlange6

**Comment by Dr15Jones** [comment]
So the LZMA change is in ROOT 6.06_patches branch: https://github.com/root-mirror/root/blob/v6-06-00-patches/core/lzma/src/ZipLZMA.c

**Comment by Dr15Jones** [comment]
The window where the first problem arrose is around 350 events into the file
```
Begin processing the 346th record. Run 276242, Event 2473989083, LumiSection 1319 at 05-Oct-2016 21:25:37.721 CEST
Begin processing the 347th record. Run 276242, Event 2475325291, LumiSection 1319 at 05-Oct-2016 21:25:39.149 CEST
Begin processing the 348th record. Run 276242, Event 2474362843, LumiSection 1319 at 05-Oct-2016 21:27:20.158 CEST
Begin processing the 349th record. Run 276242, Event 2473748478, LumiSection 1319 at 05-Oct-2016 21:27:20.162 CEST
```

**Comment by Dr15Jones** [comment]
@smuzaffar would it be possible to get a one off build of CMSSW_8_0_20_patch1 but with the version of ROOT from CMSSW_8_1_X? That would let us test the effect of the LZMA patch. We also wouldn't have ot have the most latest ROOT if we could get a version that we were sure had the LZMA patch. 

This could be used to assess the affect on the threading efficiency for this job.

**Comment by Dr15Jones** [comment]
@bbockelm what about a double buffering system? When we hit the 'cluster' limit we switch to different baskets and asynchronously handle the compression and writing to disk?

**Comment by Dr15Jones** [comment]
@pcanal So you can follow the discussion.

**Comment by Dr15Jones** [comment]
Here is the pull request to ROOT which patches lzma
https://github.com/root-mirror/root/commit/93e1f65e23c4f88316f6a5b63417e02e183f1ba2

**Comment by Dr15Jones** [comment]
#16111 fixes the incorrect reporting in edmStreamStallGrapher.py which is in CMSSW_8_0_X now.

**Comment by pcanal** [comment]
> Digging through the call graph, it seems that TBranch::FlushBaskets is mostly thread-safe:

Note that one of the feature of 6.08 is to also for a multi-thread version of TTree::GetEntry so some of the missing part have been taking care of, but likely not all of them.

>what about a double buffering system? When we hit the 'cluster' limit we switch to different 
> baskets and asynchronously handle the compression and writing to disk?

Indeed one cost of doing the compression in parallel would that each branch would now need its own 'compression' buffer .. hence increase the memory.

I still find odd that only '4' of out of the 12 or cluster are taking a very long time ... it might be worth running a profiler to get a better idea ... especially on a single thread example where it might be clearer.

**Comment by Dr15Jones** [comment]
I tested a job where I changed AODoutput to use a split level of 0. Using the improved edmStreamStallGrapher gives
```
Stalled Module                       Tot Stall Time  Stall Times
AODoutput                           : 793.47         51.08, 50.80, 48.52, 48.43, 45.27, 44.95, 40.97, 39.94, 39.75, 22.33, 21.78, 21.33, 21.08, 20.26, 20.17, 18.14, 17.35, 17.12, 16.84, 16.80, 15.18, 15.01, 14.46, 14.14, 13.30, 12.84, 12.51, 12.36, 11.07, 9.67, 9.10, 7.62, 4.19, 3.57, 1.26, 1.06, 0.95, 0.87, 0.86, 0.77, 0.76, 0.67, 0.58, 0.55, 0.53, 0.49, 0.48, 0.48, 0.45, 0.39, 0.36, 0.35, 0.35, 0.33, 0.33, 0.31, 0.29, 0.24, 0.23, 0.20, 0.20, 0.18, 0.18, 0.17, 0.17, 0.16, 0.12, 0.11, 0.11
MINIAODoutput                       : 31.80          9.59, 9.07, 6.75, 1.05, 0.87, 0.70, 0.51, 0.40, 0.38, 0.30, 0.29, 0.24, 0.18, 0.18, 0.18, 0.17, 0.15, 0.14, 0.12, 0.11, 0.11, 0.11, 0.10, 0.10
sourceFindEvent                     : 11.27          0.22, 0.20, 0.17, 0.16, 0.16, 0.16, 0.15, 0.15, 0.15, 0.14, 0.14, 0.14, 0.14, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10
valHcalTriggerPrimitiveDigis        : 3.51           1.76, 1.75
source                              : 2.23           0.11, 0.11, 0.11, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10
ALCARECOStreamEcalESAlign           : 2.02           0.57, 0.44, 0.41, 0.32, 0.28
valGctDigis                         : 1.88           1.69, 0.19
MuIsoTracks                         : 1.75           1.75
valRpcTriggerDigis                  : 1.70           1.70
l1tCsctf                            : 1.64           1.64
ALCARECOStreamHcalCalIterativePhiSym: 0.82           0.46, 0.36
initialStepTracks                   : 0.67           0.67
caloMet                             : 0.47           0.47
ckfOutInTracksFromConversions       : 0.32           0.32
calomuons                           : 0.29           0.29
ALCARECOStreamEcalUncalWElectron    : 0.21           0.21
l1compareforstage1                  : 0.19           0.19
valCsctfTrackDigis                  : 0.19           0.19
pixelLessStepClassifier1            : 0.17           0.17
fixedGridRhoFastjetCentralCalo      : 0.14           0.14
egmGedGsfElectronPFPileUpIsolation  : 0.12           0.12
patPFMetT1T2Corr                    : 0.12           0.12
pixelPairStepTrackCandidates        : 0.10           0.10
muonSeededSeedsOutInDisplaced       : 0.10           0.10
```
where the time report is
```
TimeReport> Time report complete in 2826.37 seconds
 Time Summary: 
 - Min event:   2.93923
 - Max event:   74.1427
 - Avg event:   10.2468
 - Total loop:  2705.43
 - Total job:   2826.37
 Event Throughput: 0.369627 ev/s
 CPU Summary: 
 - Total loop:  9529.66
 - Total job:   9649.13
```
Which gives an efficiency of 
```
9529.66/2705.43/4 = 0.880604931564
```

**Comment by bbockelm** [comment]
@Dr15Jones - LZMA is (purposely) slow.  Could you try zlib level 1 on a normally-split file?

@pcanal - I have been digging in "IMT" mode tonight.  It seems that we already have a compression buffer per branch when this mode is enabled!  This is indeed overkill: only a thread-local compression buffer should be needed.  I'll try to see if I can get parallel processing working based on what's already done with `GetEntry`.

**Comment by Dr15Jones** [comment]
With split level 0 the stall graph is
https://dl.dropboxusercontent.com/u/11356841/rereco_stall_split0_1000.pdf

**Comment by Dr15Jones** [comment]
So I ran with zlib. The stall information is
```
Stalled Module                       Tot Stall Time  Stall Times
MINIAODoutput                       : 36.63          12.62, 12.36, 8.71, 0.43, 0.37, 0.28, 0.24, 0.21, 0.20, 0.19, 0.17, 0.16, 0.13, 0.12, 0.11, 0.11, 0.11, 0.11
sourceFindEvent                     : 10.65          0.19, 0.17, 0.16, 0.16, 0.16, 0.16, 0.16, 0.16, 0.15, 0.15, 0.15, 0.14, 0.14, 0.14, 0.14, 0.14, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.13, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.12, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.11, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10
valHcalTriggerPrimitiveDigis        : 4.97           1.66, 1.66, 1.65
AODoutput                           : 4.00           1.11, 0.87, 0.76, 0.32, 0.26, 0.23, 0.19, 0.15, 0.11
l1tCsctf                            : 2.86           1.57, 1.29
ALCARECOStreamHcalCalIterativePhiSym: 1.95           0.79, 0.59, 0.34, 0.23
valRctDigis                         : 1.56           1.56
source                              : 1.24           0.11, 0.11, 0.11, 0.11, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10, 0.10
ckfOutInTracksFromConversions       : 0.82           0.82
detachedTripletStepTracks           : 0.55           0.55
SiPixelHitEfficiencySource          : 0.51           0.51
trackerAnalyzer                     : 0.21           0.21
SiStripMonitorTrackIB               : 0.15           0.15
valGctDigis                         : 0.12           0.12
valDttfDigis                        : 0.12           0.12
valRpcTriggerDigis                  : 0.11           0.11
mixedTripletStepSeedLayersB         : 0.11           0.11
DQMoutput                           : 0.10           0.10
```
The timing report is
```
TimeReport> Time report complete in 2439.02 seconds
 Time Summary: 
 - Min event:   2.87289
 - Max event:   30.5748
 - Avg event:   8.93697
 - Total loop:  2376.84
 - Total job:   2439.02
 Event Throughput: 0.420726 ev/s
 CPU Summary: 
 - Total loop:  9014.64
 - Total job:   9073.62
```
Where the efficiency is
```
9014.64/2376.84/4 = 0.948174887666
```

**Comment by dan131riley** [comment]
Unfortunately, subbing in a different version of root isn't as simple as a "scram setup" because of downstream dependencies, so it looks like it would take a special build to get a comparison with the lzma patches.

wrt gzip vs. lzma, on my test system at Cornell, with a local copy of the input file, processing 1000 events, vanilla setup gives

```
Stalled Module                       Tot Stall Time  Stall Times
AODoutput                           : 1349.71        68.53, 67.66, 67.42, 66.20, 66.11, 65.28, 64.70, 64.55, 64.28, 64.27, 63.99, 63.54, 63.35, 62.90, 58.15, 49.08, 47.23, 47.23, 36.07, 32.78, 28.73, 22.64, 21.83, 21.41, 11.85, 10.21, 7.67, 6.22, 5.48, 5.33, 5.05, 3.12, 1.48, 1.39, 1.23, 1.11, 1.07, 0.96, 0.94, 0.83, 0.80, 0.75, 0.72, 0.61, 0.55, 0.53, 0.39, 0.35, 0.33, 0.25, 0.23, 0.22, 0.22, 0.22, 0.20, 0.19, 0.18, 0.17, 0.17, 0.16, 0.16, 0.13, 0.11, 0.10, 0.10
MINIAODoutput                       : 14.40          6.28, 5.36, 0.30, 0.25, 0.22, 0.20, 0.20, 0.19, 0.19, 0.18, 0.15, 0.15, 0.15, 0.14, 0.12, 0.12, 0.10, 0.10
valHcalTriggerPrimitiveDigis        : 5.40           3.11, 1.15, 1.14
QGTagger                            : 3.93           1.97, 1.96
ALCARECOStreamHcalCalIterativePhiSym: 2.48           0.74, 0.66, 0.51, 0.30, 0.15, 0.12
l1tCsctf                            : 1.94           1.94
valDttfDigis                        : 1.75           1.75
valRctDigis                         : 1.74           1.74
ALCARECOStreamEcalESAlign           : 1.42           0.42, 0.28, 0.20, 0.19, 0.19, 0.14
sourceFindEvent                     : 1.05           0.36, 0.13, 0.13, 0.12, 0.11, 0.10, 0.10
valGmtDigis                         : 0.59           0.30, 0.29
hcalphisymHLT                       : 0.51           0.17, 0.17, 0.17
valGctDigis                         : 0.49           0.29, 0.20
SiPixelHitEfficiencySource          : 0.33           0.33
initialStepTracks                   : 0.21           0.21
valRpcTriggerDigis                  : 0.20           0.20
```

and with gzip instead of lzma for all outputs

```
Stalled Module                       Tot Stall Time  Stall Times
valHcalTriggerPrimitiveDigis        : 3.48           1.21, 1.14, 1.13
AODoutput                           : 1.89           0.54, 0.53, 0.36, 0.24, 0.11, 0.11
ALCARECOStreamHcalCalIterativePhiSym: 1.89           0.59, 0.47, 0.43, 0.40
valRpcTriggerDigis                  : 0.94           0.94
valRctDigis                         : 0.93           0.93
l1tCsctf                            : 0.90           0.90
sourceFindEvent                     : 0.43           0.12, 0.11, 0.10, 0.10
pixelLessStepTrackCandidates        : 0.31           0.31
ALCARECOStreamEcalESAlign           : 0.30           0.18, 0.12
lowPtTripletStepTracks              : 0.15           0.15
valCsctfTrackDigis                  : 0.10           0.10
valGctDigis                         : 0.10           0.10
valGmtDigis                         : 0.10           0.10
```

stall graph here:

https://dl.dropboxusercontent.com/u/28012153/rereco-stall-gzip-1000.pdf

So this does look like lzma, and it will be interesting to see what this looks like with the lzma memory reduction fixes.

**Comment by bbockelm** [comment]
We won't likely be able to backport it here, but this is an implementation of the parallel version of `FlushBaskets` as described above: https://github.com/root-mirror/root/pull/277

**Comment by smuzaffar** [comment]
@Dr15Jones , should we try to build a special release with CMSSW_8_0_20_patch1 code and root from 81X? Do you want it available for deployment or just local tests on cmsdevNN machines?

**Comment by Dr15Jones** [comment]
@smuzaffar local testing is all we need since we would not deploy it without validation.

**Comment by Dr15Jones** [comment]
@dan131riley Nice measurement :). Did you also convert `ALCARECOStreamHcalCalIterativePhiSym` and `ALCARECOStreamEcalESAlign` which are also OutputModules? [Assuming they were not already using gzip]

**Comment by smuzaffar** [comment]
@Dr15Jones , release is now available on vocms0301. In order to create a dev area based on it please run
```
scram project /var/build/muz/8020_patch1/cms/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_0_20_P1_NEWROOT
```

**Comment by Dr15Jones** [comment]
So I tried running the job in that release but I get the exception
```
----- Begin Fatal Exception 06-Oct-2016 23:56:52 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Processing run: 276242
   [1] Running path 'dqmoffline_step'
   [2] Calling beginRun for module SiStripMonitorTrack/'SiStripMonitorTrackCommon'
Exception Message:
MissingParameter: ParameterSet 'TH1ClusterPos' not found.
----- End Fatal Exception -------------------------------------------------
```

Looking I just realized that the release I was using was CMSSW_8_0_20 but the release I requested built was for CMSSW_8_0_20_patch1. Maybe there is a subtle difference between the two?

**Comment by davidlange6** [comment]
Right- When you have full psets most releases are incompatible in some way



On Oct 7, 2016, at 2:03 AM, Chris Jones <notifications@github.com<mailto:notifications@github.com>> wrote:


So I tried running the job in that release but I get the exception

----- Begin Fatal Exception 06-Oct-2016 23:56:52 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Processing run: 276242
   [1] Running path 'dqmoffline_step'
   [2] Calling beginRun for module SiStripMonitorTrack/'SiStripMonitorTrackCommon'
Exception Message:
MissingParameter: ParameterSet 'TH1ClusterPos' not found.
----- End Fatal Exception -------------------------------------------------


Looking I just realized that the release I was using was CMSSW_8_0_20 but the release I requested built was for CMSSW_8_0_20_patch1. Maybe there is a subtle difference between the two?

**Comment by Dr15Jones** [comment]
I was able to re-inject the values that were missing to get the job to run. I'm now running CMSS_8_0_20_patch1 on the same machine.

**Comment by Dr15Jones** [comment]
The results for using the ROOT from CMSSW_8_1_X with 8_0_20_patch1
```
TimeReport> Time report complete in 1946.69 seconds
 Time Summary: 
 - Min event:   2.26595
 - Max event:   23.0119
 - Avg event:   7.25001
 - Total loop:  1916.82
 - Total job:   1946.69
 Event Throughput: 0.521697 ev/s
 CPU Summary: 
 - Total loop:  7253.42
 - Total job:   7282.2
```
Which gets an efficiency of  `7253.42/1916.82/4 = 0.946'

The stall information is
```
Stalled Module                       Tot Stall Time  Stall Times
AODoutput                           : 69.84          5.71, 5.51, 5.18, 5.17, 5.09, 5.02, 4.88, 3.80, 3.78, 3.09, 2.97, 2.83, 2.13, 2.06, 1.81, 1.65, 1.48, 1.05, 1.03, 0.81, 0.72, 0.64, 0.57, 0.56, 0.47, 0.42, 0.41, 0.23, 0.20, 0.19, 0.14, 0.13, 0.11
valHcalTriggerPrimitiveDigis        : 4.64           1.56, 1.54, 1.54
ALCARECOStreamHcalCalIterativePhiSym: 3.57           0.88, 0.80, 0.75, 0.45, 0.35, 0.34
l1tCsctf                            : 2.81           1.44, 1.37
valDttfDigis                        : 1.60           1.48, 0.12
pixelPairStep                       : 0.40           0.27, 0.13
sourceFindEvent                     : 0.38           0.14, 0.13, 0.11
ALCARECOStreamEcalUncalWElectron    : 0.35           0.35
jetDQMAnalyzerAk4PFUncleaned        : 0.16           0.16
l1compareforstage1                  : 0.14           0.14
valRpcTriggerDigis                  : 0.12           0.12
ALCARECOStreamEcalESAlign           : 0.12           0.12
```

**Comment by Dr15Jones** [comment]
The same machine running vanilla CMSSW_8_0_20_patch1
```
TimeReport> Time report complete in 2273.74 seconds
 Time Summary: 
 - Min event:   2.46367
 - Max event:   71.2104
 - Avg event:   8.13219
 - Total loop:  2147.72
 - Total job:   2273.74
 Event Throughput: 0.465611 ev/s
 CPU Summary: 
 - Total loop:  7549.7
 - Total job:   7674.75
```
which gives the efficiency of `7549.7/2147.72/4 = 0.88`

With stall information
```
Stalled Module                       Tot Stall Time  Stall Times
AODoutput                           : 644.15         58.72, 56.56, 56.54, 51.58, 51.28, 51.07, 50.32, 49.26, 47.01, 18.15, 16.89, 14.19, 13.30, 12.63, 9.36, 9.22, 8.51, 6.26, 6.23, 5.55, 5.04, 4.72, 4.23, 4.10, 3.97, 2.70, 2.00, 1.76, 1.69, 1.62, 1.54, 1.53, 1.36, 1.36, 1.31, 1.25, 1.21, 0.99, 0.98, 0.90, 0.88, 0.83, 0.75, 0.55, 0.42, 0.38, 0.34, 0.32, 0.31, 0.30, 0.27, 0.23, 0.23, 0.17, 0.17, 0.17, 0.17, 0.16, 0.16, 0.13, 0.12, 0.10, 0.10
MINIAODoutput                       : 14.65          7.19, 6.40, 0.26, 0.19, 0.18, 0.17, 0.16, 0.10
l1tCsctf                            : 4.22           1.45, 1.44, 1.33
valHcalTriggerPrimitiveDigis        : 1.65           1.65
CleanedPFJets                       : 1.65           1.65
MuBestPV                            : 1.65           1.65
ALCARECOStreamEcalESAlign           : 1.20           0.39, 0.29, 0.21, 0.17, 0.14
ALCARECOStreamHcalCalIterativePhiSym: 1.07           0.53, 0.42, 0.12
sourceFindEvent                     : 0.35           0.13, 0.12, 0.10
valRctDigis                         : 0.20           0.10, 0.10
mergedDuplicateTracks               : 0.20           0.20
valGctDigis                         : 0.11           0.11
source                              : 0.10           0.10
```

**Comment by Dr15Jones** [comment]
Looks like using the newest ROOT gave a factor of almost 10 for timing on the output modules.

**Comment by bbockelm** [comment]
Were both of the above tests using LZMA?

**Comment by Dr15Jones** [comment]
Both used LZMA.

**Comment by dan131riley** [comment]
That's consistent with what I saw with Slava's test job when I was working on the LZMA patch--we went from 

AODoutput                         : 687.31         48.16, 44.15, 43.82, 43.53, 40.34, 38.43, 36.51, 34.71, 33.63, 24.07, 22.80, 21.76, 20.13, 19.82, 16.83, 13.97, 13.52, 12.59, 11.35, 11.21, 10.91, 10.13, 9.39, 8.67, 8.56, 8.42, 8.42, 7.80, 7.05, 6.99, 6.69, 6.50, 5.91, 5.66, 5.17, 4.64, 3.44, 3.22, 2.60, 1.56, 0.84, 0.83, 0.51, 0.50, 0.47, 0.22, 0.20, 0.18, 0.13, 0.13, 0.12, 0.12

to

AODoutput                   : 92.66          4.90, 4.82, 4.73, 4.49, 4.44, 4.34, 4.01, 3.99, 3.97, 3.77, 3.67, 3.56, 3.45, 2.93, 2.91, 2.85, 2.68, 2.49, 2.28, 2.27, 2.14, 1.96, 1.90, 1.72, 1.52, 1.46, 1.37, 1.33, 1.26, 1.16, 0.90, 0.79, 0.62, 0.51, 0.42, 0.39, 0.29, 0.14, 0.12, 0.11

Results from 

http://www.classe.cornell.edu/~dsr/cms/stall-mt4-orig.txt
http://www.classe.cornell.edu/~dsr/cms/stall-mt4.txt

Stall graphs etc. are also in

http://www.classe.cornell.edu/~dsr/cms/

> Looks like using the newest ROOT gave a factor of almost 10 for timing on the output modules.

AFAICT, there isn't a simple relationship between the stall times and the overall module timing, what we see with stall times is a combination of the module timing plus the probability of contention.  I found it useful to compare the LZMA and gzip times--as I wrote back in July (note this for 8 threads, effect for 4 threads is smaller):

"AOD still has longer stalls than I'd like, but the LZMA/zlib MT-8 loop times, 917.05/895.323 = 1.024, looks to be saying that the stalls are small enough that the overall effect on the event loop time is pretty close to zlib.  Compared to pre-patch 1107.8/895.323 = 1.24, we should be paying a much smaller real time penalty for using LZMA for AOD.

"The stall times don't seem to scale in a simple way wrt the cpu power, nor does the total run time have a simple relationship to the stall times.  Some of that is probably memory speed vs. cpu power, but it isn't obvious--I may have to write a toy simulation to try to understand those better."

-dan (no, still haven't written that toy simulation)

**Comment by smuzaffar** [comment]
@Dr15Jones , can we close this issue now?
